src/toolkits/style_transfer/style_transfer.cpp

```diff
@@ -0,0 +1,269 @@
+  opts.erase(resnet_mlmodel_path_iter);
+
+  auto vgg_mlmodel_path_iter = opts.find("vgg_mlmodel_path");
+  if (vgg_mlmodel_path_iter == opts.end()) {
+    log_and_throw("Expected option \"vgg_mlmodel_path\" not found.");
+  }
+  const std::string vgg_mlmodel_path = vgg_mlmodel_path_iter->second;
+  opts.erase(vgg_mlmodel_path_iter);
+
+  auto num_styles_iter = opts.find("num_styles");
+  if (num_styles_iter == opts.end()) {
+    log_and_throw("Expected option \"num_styles\" not found.");
+  }
+  size_t num_styles = num_styles_iter->second;
+
+  init_options(opts);
+
+  if (read_state<flexible_type>("random_seed") == FLEX_UNDEFINED) {
+    std::random_device random_device;
+    int random_seed = static_cast<int>(random_device());
+    add_or_update_state({{"random_seed", random_seed}});
+  }
+
+  m_training_data_iterator = create_iterator(style, content, true, num_styles);
+
+  m_training_compute_context = create_compute_context();
+  if (m_training_compute_context == nullptr) {
+    log_and_throw("No neural network compute context provided");
+  }
+
+  infer_derived_options();
+
+  m_resnet_spec = init_resnet(resnet_mlmodel_path, num_styles);
+  m_vgg_spec = init_vgg_16(vgg_mlmodel_path);
+
+  float_array_map weight_params = m_resnet_spec->export_params_view();
+  float_array_map vgg_params = m_vgg_spec->export_params_view();
+
+  weight_params.insert(vgg_params.begin(), vgg_params.end());
+
+  shared_float_array st_num_styles(std::make_shared<float_scalar>(num_styles));
+
+  m_training_model = m_training_compute_context->create_style_transfer(
+      {{"st_num_styles", st_num_styles}}, weight_params);
+
+  // TODO: print table printer
+}
+
+flex_int style_transfer::get_max_iterations() const {
+  return read_state<flex_int>("max_iterations");
+}
+
+flex_int style_transfer::get_training_iterations() const {
+  return read_state<flex_int>("training_iterations");
+}
+
+flex_int style_transfer::get_num_classes() const {
+  return read_state<flex_int>("num_classes");
+}
+
+void style_transfer::perform_training_iteration() {
+  ASSERT_TRUE(m_training_data_iterator != nullptr);
+  ASSERT_TRUE(m_training_model != nullptr);
+
+  flex_int iteration_idx = get_training_iterations();
+  flex_int batch_size = read_state<flex_int>("batch_size");
+
+  std::vector<st_example> batch =
+      m_training_data_iterator->next_batch(batch_size);
+
+  for (size_t y = 0; y < batch.size(); y++) {
+    size_t style_index = batch[y].style_index;
+
+    turi::image_type content_image = batch[y].content_image;
+    turi::image_type style_image = batch[y].style_image;
+
+    shared_float_array index_array(std::make_shared<float_scalar>(style_index));
+
+    std::map<std::string, shared_float_array> results =
+        m_training_model->train({{"input", prepare_images(content_image)},
+                                 {"labels", prepare_images(style_image)},
+                                 {"index", index_array}});
```

Comment:
Doesn't this only pass one image at a time to the backend? Shouldn't we be passing an entire batch at a time?

Reply:
Good catch